### ISSUE TITLE
handling escaping and identifier naming rules in Nix

### ISSUE THREAD

**Issue by cdepillabout** [cdepillabout/cabal2nixWithoutIFD#10]
One problem in `purenix` is escaping attrset keys in Nix, as well as rules for naming identifiers.

Our big problem is that PureScript has one set of rules for what is a valid record key (and the rules for escaping), and Nix has a different set of rules.  Also, PureScript has one set of rules for what is a valid identifier, and Nix has a different set of rules.

We need to take time to investigate PureScript's rules vs. Nix's rules.  We need to make sure that `purenix` can't accidentally generate Nix code with invalid identifiers or incorrectly-escaped attrset keys.

Here's how we are thinking of handling this:

-   We we planning on basically ignoring this for the initial prototype.  When writing PureScript code, we can just be careful not to use any identifiers or attrset keys that wouldn't work.  This should be easy since we control all the code, and we have to write a custom `Prelude` anyway.
-   Before really looking into this, we need a test suite so that we can write tests to make sure the escaping rules in `purenix` make sense, and we correctly reject bad identifiers.
-   There are some identifiers that are valid in PureScript but are invalid in Nix.  For example, `assert` is a keyword in Nix, but a valid identifier in PureScript.  We need to make sure to disallow keywords like this in identifiers in Nix, and to always quote keywords in attrset keys.
-   We discussed completely banning the use of any identifier that would shadow a top-level built-in in Nix.  Some examples of top-level builtins in Nix are `map`, `false`, `fetchGit`, etc.  You can get a full list by running `nix-repl` and hitting <kbd>TAB</kbd>.

**Comment by cdepillabout** [comment]
One thing I found when playing around with this is that in Nix, it is possible to quote a let binding:

```nix
let "assert" = 3; in ...
```

The problem is that I don't know how to then use `"assert"`.

One possibility is to `inherit` it into an attrset, but this feels somewhat hacky:

```nix
> let "assert" = 3; in {inherit "assert";}."assert"
3
```

Note that it is not possible to quote function arguments:

```nix
> let x = "assert": 3; in x 100                     
error: syntax error, unexpected ':', expecting ';', at (string):1:17
```

**Comment by cdepillabout** [comment]
One other thing I noticed is what happens when constructors and functions have a single-quote (prime) in their names.

This PureScript code:

```purescript
module Main where

data MyType = MkFoo' Int

myFunc' :: MyType
myFunc' = MkFoo' 3

example :: Int
example =
  case myFunc' of
    MkFoo' i -> i
```

gets translated into this code by `purenix`:

```nix
modules: 
let
  "MkFoo'" = value0: 
    { __tag = "MkFoo'";
      __field0 = value0;
    };
  "myFunc'" = modules.Main."MkFoo'" 3;
  example = 
    let
      __pattern0 = __fail: 
        if modules.Main."myFunc'".__tag == "MkFoo'"
          then 
            let
              i = modules.Main."myFunc'".__field0;
            in
              i
          else __fail;
      __pattern1 = builtins.throw "Pattern match failure";
    in
      __pattern0 __pattern1;
in
  {inherit MkFoo' myFunc' example;}
```

There are two problems here:

1. In the last line, `MkFoo'` and `myFunc'` should be quoted so they can actually be exported.
2. Currently `MkFoo'` and `myFunc'` are used through `modules.Main.`, which should work.  However, if these two identifiers don't get exported, then this will no longer work.

-----------------------------------------------------

We might be able to generate identifiers liek `MkFoo'` as something like `MkFoo-prime` in Nix, since `-` is valid to have in identifier names in Nix, but not in PureScript.

**Comment by cdepillabout** [comment]
In https://github.com/cdepillabout/cabal2nixWithoutIFD/pull/12#discussion_r707852299 we talk about handling a couple more identifiers that we make use of internally.  We should make sure that users can't create identifiers with these names.

**Comment by cdepillabout** [comment]
Jonas fixed this up in #16.
